### PR TITLE
HPCC-18658 Refactor ESP sessionID and digital signature data types

### PR DESCRIPTION
**Pull request by RussWhitehead**
Currently ESP stores a user's digital signature and ESP Session ID in
MemoryBuffers. The ESP Session ID is really just an unsigned value,
and should be stored and returned as such. The digital signature will
be a base 64 encoded string, and should be passed as const char *.
This PR makes those changes

HPCC-18658 Refactor ESP sessionID and digital signature data types

Currently ESP stores a user's digital signature and ESP Session ID in
MemoryBuffers. The ESP Session ID is really just an unsigned value,
and should be stored and returned as such. The digital signature will
be a base 64 encoded string, and should be passed as const char *.
This PR implements those changes

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by RussWhitehead** [comment]
This PR replaces https://github.com/hpcc-systems/HPCC-Platform/pull/10596

**Comment by RussWhitehead** [comment]
@jakesmith Please rereview based on your comments on the previous PR https://github.com/hpcc-systems/HPCC-Platform/pull/10596

**Review by jakesmith — CHANGES_REQUESTED** [review]

**Comment by jakesmith** [comment]
@RussWhitehead - have only looked at the serialize / deserialize methods, but I think there are still problems.

Dali needs to continue to understand the old serialized format, otherwise it will break clients that connect to it prior to this build.
Conversely, newer clients cannot send a different serialization format to an old Dali for the same reason.

You _could_ check the server version when you serialize and only serialize the new format if the server (Dali) is new enough, but you'd still need to handle both formats at the server.

I think you're better of sticking with the current serialization format, but that doesn't stop you changing the members and methods to use a plain unsigned and string as members and as return types to e.g. querySessionToken().
i.e. instead of changing the serialization format, ensure that the plain token and string are serialized in the existing MemoryBuffer container format and then deserialize them out the other side. As long as the content remains the same between old and new, you can unpack them differently and expose them via methods without having to expose the raw MemoryBuffer's

**Comment by RussWhitehead** [comment]
@jakesmith the "old dali" message you refer to above (mb.append(sessionToken.length()).append(sessionToken).append(signature.length()).append(signature);) was something I added in a previous PR to 7.0, so will not break. However, I think this new methodology might be better served by a new message with version checking (MSR_LOOKUP_LDAP_PERMISSIONS_EX) what are your thoughts on that?

**Comment by jakesmith** [comment]
@RussWhitehead - ok, I didn't realize this change was to a change that's in master/7.0 only.
In which case, changing serializeExtra/deserializeExtra is fine and will not break any current releases.
But as you say, to be cleaner to toggle seriliazation of new info. based on dali version (as you are doing), but then yes send it as a new message type, so that the deserialziation code is cleaner.

**Comment by RussWhitehead** [comment]
@jakesmith Please review. I added a new message enum which is only used when the Dali is signature-aware and a signature is present.  Note that I have not checked in the code to compute a signature, so these changes are benign.  I test using all 4 permutations of legacy/new Dali/eclagent

File: system/security/shared/caching.cpp [hunk 1]
```diff
@@ -358,7 +358,7 @@ bool CPermissionsCache::lookup(ISecUser& sec_user)
             const char* cachedpw = user->queryUser()->credentials().getPassword();
             const char * pw = sec_user.credentials().getPassword();
 
-            if ((sec_user.credentials().getSessionToken().length() > 0) ||  (sec_user.credentials().getSignature().length() > 0))
+            if (sec_user.credentials().getSessionToken() != 0 || sec_user.credentials().getSignature())
```
Comment (jakesmith) [inline system/security/shared/caching.cpp]: >sec_user.credentials().getSignature()

wrapping with !isEmpty() looks appropriate.

**Comment by jakesmith** [comment]
@RussWhitehead - a few minor comments.

**Comment by RussWhitehead** [comment]
Closing, will address with separate Jira/PR

**Pull request closed without merging**